### PR TITLE
improve update-not-possible logging

### PR DESCRIPTION
**Pull request by jakecoffman**
When doing an update, you might get an update_not_possible error with the following logs:

```
updater | Latest version is 1.0.0
updater | Requirements to unlock update_not_possible
updater | Requirements update strategy bump_versions
updater | No update possible for my-dependency 1.0.0
```

One of the reasons this can happen is that all versions were filtered out while trying to update a specific dependency. 

This PR aims to improve the logging for ecosystems to give users a hint of why the update is not possible:

```
updater | Latest version is 1.0.0
updater | Requirements to unlock update_not_possible
updater | Requirements update strategy bump_versions
updater | Filtered out 156 unsupported Python 3.11 versions
updater | Filtered out 12 pre-release versions
updater | Filtered out 256 ignored versions
updater | No update possible for my-dependency 1.0.0
```

### CODE REVIEW

**Comment by jakecoffman** [comment]
Rather than try and test all of these new log messages, I added Sorbet types to make sure I didn't typo anything. That should be enough to verify this won't break things when deploying.

Some of the types were tricky though as a version array could be an array of Version, or a Hash of Symbols to stuff, including a Version. Rather than go down that rabbit hole, I left them as `T::Array[T.untyped]` since that's enough to verify I didn't break anything!

File: bundler/lib/dependabot/bundler/update_checker/latest_version_finder.rb
Comment (abdulapopoola) [inline bundler/lib/dependabot/bundler/update_checker/latest_version_finder.rb:86]: do we want to add a log here in case it doesn't raise? Would that help with debugging going forward too?

File: common/lib/dependabot/update_checkers/version_filters.rb
Comment (abdulapopoola) [inline common/lib/dependabot/update_checkers/version_filters.rb:20]: can these types be named? Tagging @bdragon and @JamieMagee to weigh in

File: cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb
Comment (abdulapopoola) [inline cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb:62]: just wondering; is this something that can be raised higher in the inheritance hierarchy so we don't have to implement it in every downstream implementation?

File: bundler/lib/dependabot/bundler/update_checker/latest_version_finder.rb
Comment (jakecoffman) [reply bundler/lib/dependabot/bundler/update_checker/latest_version_finder.rb:86]: I did, it's just below 👇

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (abdulapopoola) [inline gradle/lib/dependabot/gradle/update_checker/version_finder.rb:97]: [purely curious so feel free to ignore] Do you know what the 1900 means?

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (abdulapopoola) [inline gradle/lib/dependabot/gradle/update_checker/version_finder.rb:109]: not matching the type? Can that be explained some more?

File: cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb
Comment (jakecoffman) [reply cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb:62]: Yes there are a few things like this and also filtering out versions that are lower than the current version that we could potentially pull up into the common updater code.

Comment (abdulapopoola) [reply cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb:62]: And if that's prohibitively expensive or not feasible, please can we fix it for all ecosystems in one go so we don't have it linger :)

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (jakecoffman) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb:97]: If it's a date jammed into a version class then that line will filter it out? That's my guess.

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (jakecoffman) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb]: Just looked up what these suffixes are actually called, they're artifact classifiers, so I updated the log message. Thanks for catching that.

File: cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb
Comment (jakecoffman) [reply cargo/lib/dependabot/cargo/update_checker/latest_version_finder.rb:62]: It's definitely out of scope for this PR, I can file an issue to track that work.

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (jakecoffman) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb:109]: In context with the other log messages around it, I think it would look like:

```
updater | Latest version is 1.0.0-android
updater | Requirements to unlock update_not_possible
updater | Requirements update strategy bump_versions
updater | Filtered out 156 versions with a different classifier
updater | No update possible for my-dependency 1.0.0-android
```

I think it's clear in context assuming folks working in the ecosystem know it's called a classifier, but we can add it if you think that would help.

Comment (jakecoffman) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb:109]: I know the folks in Maven ecosystem know, this is how it would be specified in the pom:

```xml
<dependency>
   <groupId>org.example</groupId>
   <artifactId>stuff</artifactId>
   <version>1.0.0</version>       
   <classifier>android</classifier>
</dependency>
```

Comment (abdulapopoola) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb:109]: My 2c: the more verbose we are in logging; the more useful it is when debugging.

**Review by JoeRobich — APPROVED** [review]
NuGet changes look good to me.

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (jakecoffman) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb]: I added the classifier name that we're filtering on to the log message.

File: common/lib/dependabot/update_checkers/version_filters.rb
Comment (jakecoffman) [reply common/lib/dependabot/update_checkers/version_filters.rb:20]: Generic type parameters across many languages use T, U, V in that order for type names.

Commonly the Java docs are referenced as the convention:
- https://docs.oracle.com/javase/tutorial/java/generics/types.html

Here's an example in the Go source code:
- https://github.com/golang/go/blob/a3f69c778fa6c082d38f9cdfacab27714f08a343/src/cmd/cgo/internal/test/typeparam.go#L10

Example in the Cargo book using `<T>` and the other cases:
- https://doc.rust-lang.org/book/ch10-00-generics.html

C++ using <T>:
- https://en.cppreference.com/w/cpp/language/templates

Comment (abdulapopoola) [reply common/lib/dependabot/update_checkers/version_filters.rb:20]: Sorry I wasn't clear - I understand the T, V, U naming convention for generics (I've used it too).

I meant to ask about `type aliases` like [this](https://github.com/dependabot/dependabot-core/pull/9232#discussion_r1516598410) example.

Comment (jakecoffman) [reply common/lib/dependabot/update_checkers/version_filters.rb:20]: Ah ok, I can make type aliases for `T::Array[Gem::Version]` and `T::Array[T.untyped]` if you think that would be helpful.

Comment (abdulapopoola) [reply common/lib/dependabot/update_checkers/version_filters.rb:20]: Only if it helps with readability :) Your call.

Comment (jakecoffman) [reply common/lib/dependabot/update_checkers/version_filters.rb:20]: Worth noting, I think `T::Hash[Symbol, Gem::Version]` is actually incorrect, it's a Hash that for the key `:version` it's a `Gem::Version`, but all the other keys are not `Gem::Version` so might make sense to introduce a proper class instead which would eliminate the need for a type alias.

File: gradle/lib/dependabot/gradle/update_checker/version_finder.rb
Comment (jurre) [reply gradle/lib/dependabot/gradle/update_checker/version_finder.rb]: Yeah I think this is good, thanks! Tbh I am not sure if everyone involved knows what the classifier is, tbh, I didn't really know what it was either.

File: common/lib/dependabot/update_checkers/version_filters.rb
Comment (JamieMagee) [inline common/lib/dependabot/update_checkers/version_filters.rb:15]: nit: add a link to the documentation here https://sorbet.org/docs/generics#placing-bounds-on-generic-methods

File: common/lib/dependabot/update_checkers/version_filters.rb
Comment (jakecoffman) [reply common/lib/dependabot/update_checkers/version_filters.rb:21]: Captured this info here so we can come back to it after this PR merges: https://github.com/dependabot/dependabot-core/issues/9283

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**
